From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Fri, 24 Apr 2026 17:27:37 -0500
Subject: [PATCH 42/46] 
 0051-win32u-Implement-dead-key-support-for-MAPVK_VK_TO_CH.patch

---
 dlls/win32u/input.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 1a2a8457f02..d8655e8686f 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -441,6 +441,7 @@ static void kbd_tables_init_vk2char( const KBDTABLES *tables, UINT vk2char[0x100
 {
     const VK_TO_WCHAR_TABLE *table;
     const VK_TO_WCHARS1 *entry;
+    UINT is_dead = 0;
 
     memset( vk2char, 0, 0x100 * sizeof(*vk2char) );
 
@@ -448,9 +449,26 @@ static void kbd_tables_init_vk2char( const KBDTABLES *tables, UINT vk2char[0x100
     {
         for (entry = table->pVkToWchars; entry->VirtualKey; entry = NEXT_ENTRY(table, entry))
         {
+            UINT temp = is_dead;
+
+            /* if the dead key we inherited
+             * from the previous iteration fails the below conditions
+             * then reset the is_dead flag */
+            is_dead = 0;
+
             if (entry->VirtualKey & ~0xff) continue;
             if (entry->wch[0] == WCH_NONE) continue;
-            vk2char[entry->VirtualKey] = entry->wch[0];
+            /* A..Z is already handled and the others are unshifted */
+            if (entry->Attributes & SGCAPS) continue;
+            /* avoid setting dead bit with null wch */
+            if (!entry->wch[0]) continue;
+            if (entry->wch[0] == WCH_DEAD)
+            {
+                is_dead |= (1u << 31);
+                continue;
+            }
+
+            vk2char[entry->VirtualKey] = entry->wch[0] | temp;
         }
     }
 }
